### PR TITLE
Support Poisson regression

### LINKED ISSUE #233: poisson regresson in lightgbm
Thank you really very much for your amazing package! 
Any plans to introduce:
a. poisson regression as objective?
b. Offset parameter like in poisson regression?

This would permit to handle variable exposure observations in count regressions

### PR DESCRIPTION
**Pull request by henry0312**
Close #233

### CODE REVIEW

**Comment by henry0312** [comment]
sample script:

```python
# coding: utf-8
# pylint: disable = invalid-name, C0111
import lightgbm as lgb
import pandas as pd
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import GridSearchCV

# load or create your dataset
print('Load data...')
df_train = pd.read_csv('../regression/regression.train', header=None, sep='\t')
df_test = pd.read_csv('../regression/regression.test', header=None, sep='\t')

y_train = df_train[0]
y_test = df_test[0]
X_train = df_train.drop(0, axis=1)
X_test = df_test.drop(0, axis=1)

# train
gbm = lgb.LGBMRegressor(objective='poisson',
                        num_leaves=31,
                        learning_rate=0.05,
                        n_estimators=200)
gbm.fit(X_train, y_train,
        eval_set=[(X_test, y_test)],
        eval_metric='poisson',
        early_stopping_rounds=5)
```

```sh
❯ python test_sklearn.py
Load data...
[1]	valid_0's poisson: 1.78052
Train until valid scores didn't improve in 5 rounds.
[2]	valid_0's poisson: 1.46034
[3]	valid_0's poisson: 1.30203
[4]	valid_0's poisson: 1.19858
[5]	valid_0's poisson: 1.12772
[6]	valid_0's poisson: 1.07343
[7]	valid_0's poisson: 1.03415
[8]	valid_0's poisson: 1.00191
[9]	valid_0's poisson: 0.976774
[10]	valid_0's poisson: 0.955738
<snip...>
[100]	valid_0's poisson: 0.792115
[101]	valid_0's poisson: 0.791944
[102]	valid_0's poisson: 0.792073
[103]	valid_0's poisson: 0.791845
[104]	valid_0's poisson: 0.791726
[105]	valid_0's poisson: 0.791678
[106]	valid_0's poisson: 0.791766
[107]	valid_0's poisson: 0.79145
[108]	valid_0's poisson: 0.791798
[109]	valid_0's poisson: 0.791885
[110]	valid_0's poisson: 0.791783
[111]	valid_0's poisson: 0.791846
[112]	valid_0's poisson: 0.791886
Early stopping, best iteration is:
[107]	valid_0's poisson: 0.79145
```

File: src/metric/regression_metric.hpp [hunk 1]
```diff
@@ -162,5 +162,23 @@ class FairLossMetric: public RegressionMetric<FairLossMetric> {
...
+/*! \brief Poisson regression loss for regression task */
+class PoissonMetric: public RegressionMetric<PoissonMetric> {
+public:
+  explicit PoissonMetric(const MetricConfig& config) :RegressionMetric<PoissonMetric>(config) {
+  }
+
+  inline static double LossOnPoint(float label, double score, double, double) {
+    const double eps = 1e-10f;
+    if (score < eps) {
+      score = eps;
```
Comment (guolinke) [inline src/metric/regression_metric.hpp]: just a confirm: if score < 0, will assign to eps ?

Comment (henry0312) [inline src/metric/regression_metric.hpp]: yes. this is because score should be positive.
cf. score is:
image
(from https://en.wikipedia.org/wiki/Poisson_regression)

File: docs/Python-API.md [hunk 2]
```diff
@@ -652,6 +652,8 @@ The methods of each Class is in alphabetical order.
         It is used to control the width of Gaussian function to approximate hessian.
     fair_c : float
         Only used in regression. Parameter for Fair loss function.
+    poisson_max_delta_step : float
+        parameter used to safeguard optimization in Poisson regression.
```
Comment (wxchan) [inline docs/Python-API.md]: add it and gaussian_eta, fair_c to L595(`__init__ `) too

File: src/metric/regression_metric.hpp (hunk 1 above)
Comment (Laurae2) [inline src/metric/regression_metric.hpp]: You can directly return 0 for this I think. As long as it doesn't affect the hessian it should be all good.

Comment (henry0312) [inline src/metric/regression_metric.hpp]: I thinks it's better to return `eps` because of e^{theta*x} > 0

Comment (Laurae2) [inline src/metric/regression_metric.hpp]: Ah yes that's true. Myself I was using eps in one of my packages for Poisson regression...
